File: test-support/src/test/java/com/rafaelfiume/prosciutto/test/StubbedServerTest.java
package com.rafaelfiume.prosciutto.test;

import org.apache.commons.io.IOUtils;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.URL;

import static com.rafaelfiume.prosciutto.test.SalumeApiContractExampleReader.supplierAdviceForExpertRequest;
import static com.rafaelfiume.prosciutto.test.SalumeApiContractExampleReader.supplierAdviceForExpertResponse;
import static org.hamcrest.Matchers.containsString;
import static org.junit.Assert.assertThat;

public class StubbedServerTest {

    @Rule
    public DependsOnServerRunningRule server = new DependsOnServerRunningRule();

    @Rule
    public ExpectedException exception = ExpectedException.none();

    @Test
    public void shouldReturnPrimedSuccessfulResponse() throws IOException {
        whenPrimingSupplierResponseWith(supplierAdviceForExpertResponse());
        assertThat(get(supplierAdviceForExpertRequest()), containsString("<product-advisor>"));
    }

    @Test
    public void shouldThrowPrimedException() throws IOException {
        whenPrimingServerErrorWhenRequesting("/I/would/like/to/buy/bananas");

        exception.expect(RuntimeException.class);
        exception.expectMessage("Response code for url (http://localhost:8081/I/would/like/to/buy/bananas) is: 500");

        get("http://localhost:8081/I/would/like/to/buy/bananas");
    }

    private void whenPrimingSupplierResponseWith(String response) {
        server.primeSuccessfulResponse("/salume/supplier/advise/for/Expert", response);
    }

    private void whenPrimingServerErrorWhenRequesting(String response) {
        server.primeServerErrorWhenRequesting(response);
    }

    // Duplicated :( See integration package
    private static String get(String url) throws IOException {
        HttpURLConnection http = null;
        try {
            http = (HttpURLConnection) new URL(url).openConnection();

        if (http.getResponseCode() != 200) {
            // Replace by ConnectedException
            throw new RuntimeException(String.format(
                    "Response code for url (%s) is: %s", url, http.getResponseCode()));
        }

            return IOUtils.toString(http.getInputStream());
        } finally {
            http.disconnect();
        }
    }
}

File: test-support/src/test/kotlin/com/rafaelfiume/prosciutto/test/StubbedServerTest.kt
package com.rafaelfiume.prosciutto.test

import org.apache.commons.io.IOUtils
import org.junit.Rule
import org.junit.Test
import org.junit.rules.ExpectedException

import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL

import com.rafaelfiume.prosciutto.test.SalumeApiContractExampleReader.supplierAdviceForExpertRequest
import com.rafaelfiume.prosciutto.test.SalumeApiContractExampleReader.supplierAdviceForExpertResponse
import org.hamcrest.Matchers.containsString
import org.junit.Assert.assertThat

class StubbedServerTest {

    @Rule @JvmField
    var server = DependsOnServerRunningRule()

    @Rule @JvmField
    var exception = ExpectedException.none()

    @Test
    @Throws(IOException::class)
    fun shouldReturnPrimedSuccessfulResponse() {
        whenPrimingSupplierResponseWith(supplierAdviceForExpertResponse())
        assertThat(get(supplierAdviceForExpertRequest()), containsString("<product-advisor>"))
    }

    @Test
    @Throws(IOException::class)
    fun shouldThrowPrimedException() {
        whenPrimingServerErrorWhenRequesting("/I/would/like/to/buy/bananas")

        exception.expect(RuntimeException::class.java)
        exception.expectMessage("Response code for url <<http://localhost:8081/I/would/like/to/buy/bananas>> is: 500")

        get("http://localhost:8081/I/would/like/to/buy/bananas")
    }

    private fun whenPrimingSupplierResponseWith(response: String) {
        server.primeSuccessfulResponse("/salume/supplier/advise/for/Expert", response)
    }

    private fun whenPrimingServerErrorWhenRequesting(response: String) {
        server.primeServerErrorWhenRequesting(response)
    }

    // TODO Duplicated :( See integration package
    @Throws(IOException::class)
    private operator fun get(url: String): String {
        var http: HttpURLConnection? = null
        try {
            http = URL(url).openConnection() as HttpURLConnection

            if (http.responseCode != 200) {
                throw RuntimeException("Response code for url <<$url>> is: ${http.responseCode}")
            }

            return IOUtils.toString(http.inputStream)
            
        } finally {
            http!!.disconnect()
        }
    }
}
